src/plots/cartesian/axes.js

```diff
@@ -776,25 +781,44 @@ axes.calcTicks = function calcTicks(ax, opts) {
 
             var A = tickVals[a].value;
             var B = tickVals[b].value;
+            var actualDelta = Math.abs(B - A);
+            var delta = definedDelta || actualDelta;
+            var periodLength = 0;
 
-            var delta = definedDelta || Math.abs(B - A);
-            if(delta >= ONEDAY * 365) { // Years could have days less than ONEAVGYEAR period
-                v += ONEAVGYEAR / 2;
-            } else if(delta >= ONEAVGQUARTER) {
-                v += ONEAVGQUARTER / 2;
-            } else if(delta >= ONEMINMONTH) { // Months could have days less than ONEAVGMONTH period
-                var actualDelta = Math.abs(B - A);
+            if(delta >= ONEMINYEAR) {
+                if(actualDelta >= ONEMINYEAR && actualDelta <= ONEMAXYEAR) {
+                    periodLength = actualDelta;
+                } else {
+                    periodLength = ONEAVGYEAR;
+                }
+            } else if(definedDelta === ONEAVGQUARTER && delta >= ONEMINQUARTER) {
+                if(actualDelta >= ONEMINQUARTER && actualDelta <= ONEMAXQUARTER) {
+                    periodLength = actualDelta;
```

Comment:
Do we need an `else periodLength = ONEAVGQUARTER` here, to cover the case of quarter tickformat but longer `dtick`?

Reply:
That's the case I thought it could help.

Reply:
At first I thought we need an else blocks for weeks and quarters to make them similar to others.
But ( as you mentioned) since we don't generate auto labels for weeks and quarters that could lead to the problem of moving daily/monthly labels by a week/quarter. So those else blocks were removed in https://github.com/plotly/plotly.js/pull/5089/commits/beed352265254154e165ade9a89073d9564d8677 commit.

Reply:
> Here's the case I think this `else` would address: quarter period ticklabels, and dtick of 6 months or 1 year https://codepen.io/alexcjohnson/pen/abNBYxZ?editors=0010
> Currently this puts the quarter label right on the tick:
> <img alt="Screen Shot 2020-08-21 at 11 53 23 AM" width="672" src="https://user-images.githubusercontent.com/2678795/90910266-09198d80-e3a5-11ea-96c0-50db6439fc58.png">
> but just like when we have monthly labels and 3-month dtick and we shift the label half a month over anyway, we should still be shifting the label half a quarter in this case. 

Good catch. This one fixed by https://github.com/plotly/plotly.js/pull/5089/commits/350a10acce19a14084af9c5d3af2979916ffab37.
Here is an [updated codepen](https://codepen.io/MojtabaSamimi/pen/JjXyKgw?editors=0010).

Reply:
> Note: if you turn rangebreaks back on in that codepen (notice I renamed to rangebreaksx to disable them) and disable the explicit dtick, then all manner of bad things happen as you zoom and pan - unpredictable and nonuniform tick spacing, strange jumps...

Now I also put `rangebreaks` back in the above pen; and it looks OK.